build(aio): add authoring dgeni package

petebacondarwin opened the pull request:
This task is suitable for day to day docs authoring.

This task cuts corners, which makes it much faster than a full `yarn docs`
run but it does not produce completely valid output.
In general this isgood enough for authors to see their changes as they make them.

The task is triggered by a call to

```
yarn docs-watch
```

This sets up watchers on the `aio/contents` and `packages` folders.
Any changes to files below these folders new doc generation run to start.
The input to the generation is confined to a collection of files related
to the changed file. For example:

* a change to a file in `aio/content/marketing` will generate all the
marketing files.
* a change to a file in `aio/content/tutorial` or `aio/examples/toh-*`
will generate all the tutorial files (and their embedded examples).
* a change to a file in `aio/guide` or `aio/examples` (but not a `toh-`
example) will generate the appropriate guide and its embedded examples
* a change to a file in `packages` or `packages/examples` will generate
the appropriate API doc and its embedded examples.

Be aware that the mapping between docs and its examples are based on doc file
and example folder structure being equivalent. Sometimes a doc will reference
an example in a different folder, in which case the generated doc will be
inaccurate. Mostly this is not a big problem.

mary-poppins commented:
The angular.io preview for 4e1bcc8 is available [here][1].

[1]: https://pr15927-4e1bcc8ff91073f0f61b0c7cbebfbc136419f20a.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5ae9db5 is available [here][1].

[1]: https://pr15927-5ae9db5031c100db2e335939061075208b0e8f6d.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5c3f67c is available [here][1].

[1]: https://pr15927-5c3f67c8429015d226301b2a270db395a6af0506.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 55f8bfd is available [here][1].

[1]: https://pr15927-55f8bfdcae91dd3aee1c1ddde4ac7f379650cd72.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 32725f5 is available [here][1].

[1]: https://pr15927-32725f5885a42d8ba7ccbd45ac01ad3394525eed.ngbuilds.io/

Foxandxss reviewed (commented):
This work excelent except that HTTP chapter.

Should we rename it or fix it here?

mary-poppins commented:
The angular.io preview for f6e8d9d is available [here][1].

[1]: https://pr15927-f6e8d9d755badf1faca3dae4fd1fad8f2e4fb04c.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 22819f5 is available [here][1].

[1]: https://pr15927-22819f5bdb801043ab1c645b68753278793d71c9.ngbuilds.io/

petebacondarwin commented:
I think we should rename the `http` example to match the guide.
I have added tests and a basic README to this PR.

wardbell commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
 
-Run `ng e2e` to execute the end-to-end tests via [Protractor](http://www.protractortest.org/).
-Before running the tests make sure you are serving the app via `ng serve`.
+## Guide to Authoring
```
## Guide to authoring

wardbell reviewed (changes requested):
**Well done!**
I'm only commenting on the README and the author experience. I'm pleased with both!

The changes to the README are small enough to make and then lets merge this so we can see how it really feels under the pressure of active editing.

wardbell commented:
I am in favor of fixing all mismatch example folders to match their corresponding doc folders

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
-Run `ng github-pages:deploy` to deploy to GitHub Pages.
+* **API docs**: descriptions of the modules, classes, interfaces, decorators, etc that make up the Angular platform.
+API docs are generated directly from the source code.
+The source code is contained in TypeScript files, located in the `/packages` folder.
```
I'm going to prefix all the file paths with `angular/` to make it more clear (hopefully) :-)

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
+* **API docs**: descriptions of the modules, classes, interfaces, decorators, etc that make up the Angular platform.
+API docs are generated directly from the source code.
+The source code is contained in TypeScript files, located in the `/packages` folder.
+Each API item may have a preceding comment, which contains JSDoc style tags and content. The actual text is written in markdown.
```
Yes, we need to fill this out.
Interestingly even the guides, etc are not "just" markdown :-) Really they look like:

```
Some optional content block in markdown
@title
The title of the doc in markdown
@description
The content of the document in markdown
```

Anything before any jsdoc tag is combined with anything following the `@description` tag and call the "content" of the document.

It just turns out that if you do not care to provide a title then you can get away with

```
Just some content in markdown
```

petebacondarwin replied:
but let's write all of that up in another PR (and document)

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
+Each API item may have a preceding comment, which contains JSDoc style tags and content. The actual text is written in markdown.
 
-## Further help
+* **Other content**: guides, tutorials and other marketing material
```
fixed

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
 
-## Further help
+* **Other content**: guides, tutorials and other marketing material
+All other content is written using markdown in text files, located in the `/aio/content` folder.
```
See https://github.com/angular/angular/pull/15927#discussion_r111690205

So it turns out that HTML is just a subset of markdown, which means that the following files will render the same:

**a.html**

```html
<h2>Section heading</h2>
<p>Some content</p>
```

**a.md**
```
## Section heading

Some content
```

**a2.md**
```
@description
## Section heading

Some content
```

petebacondarwin replied:
I would be OK to call them all `.md` but I was led to believe that the marketing people wanted to have them as `.html`.
Also I think that it is useful that only the `marketing` folder will parse files that end in `.html`, which helps to encourage a mindset of the authors that HTML is not really allowed in the guides in general. If we start allowing people to call their files `.html` then there would be a proliferation of HTML styling in the docs.
We agreed that marketing docs are "different" in this regard.

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
+The main task for generating the docs is `yarn docs`. This will process all the source files (API and other),
+extracting the documentation and generating JSON files that can be consumed by the web app.
+
+Since this task can take up to 1 minute to run, it is not very helpful during day to day document authoring,
```
updated

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
+
+Instead you can use the `yarn docs-watch` task to watch the source files and only process the files necessary
+to generate the docs that are related to the file that has changed. Since this task takes shortcuts, it is
+much fast (often less than 1 second) but is not guaranteed to produce completely accurate content. This is
```
updated

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
+
+The general setup is as follows:
+
+* Open a terminal, ensure the dependencies are installed; run an initial doc generation; then start the web app:
```
OK

petebacondarwin commented on aio/README.md:
````diff
@@ -1,31 +1,79 @@
+* Open a second terminal and start watching the docs
+
+```bash
+yarn docs-watch
````
You can pass the `-o` flag through to the `ng serve` command by using the double dash approach:

```bash
yarn start -- -o
```

The Angular CLI does not currently support opening a specific URL.

petebacondarwin replied:
added a note about that

petebacondarwin commented on aio/README.md:
```diff
@@ -1,31 +1,79 @@
+* Open a browser at https://localhost:4200/ and navigate to the document on which you want to work.
+
+* Make changes to the page's associated doc or example files. Every time a file is saved, the doc will
+be regenerated, the app will rebuild and the page will reload.
```
Right now, the webpack dev server watches every file below the `src` folder. So when we change one of those files, it triggers webpack to rebuild, which then triggers live-reload to refresh the browser. Not sure what we can do to improve that without getting access to the innards of the webpack config.

wardbell replied:
Not killing us at the moment. Let's push forward.

petebacondarwin committed:
build(aio): add `doc-watch` yarn task for docs authors

This task is suitable for day to day docs authoring.

This task cuts corners, which makes it much faster than a full `yarn docs`
run but it does not produce completely valid output.
In general this isgood enough for authors to see their changes as they make them.

The task is triggered by a call to

```
yarn docs-watch
```

This sets up watchers on the `aio/contents` and `packages` folders.
Any changes to files below these folders new doc generation run to start.
The input to the generation is confined to a collection of files related
to the changed file. For example:

* a change to a file in `aio/content/marketing` will generate all the
marketing files.
* a change to a file in `aio/content/tutorial` or `aio/examples/toh-*`
will generate all the tutorial files (and their embedded examples).
* a change to a file in `aio/guide` or `aio/examples` (but not a `toh-`
example) will generate the appropriate guide and its embedded examples
* a change to a file in `packages` or `packages/examples` will generate
the appropriate API doc and its embedded examples.

Be aware that the mapping between docs and its examples are based on doc file
and example folder structure being equivalent. Sometimes a doc will reference
an example in a different folder, in which case the generated doc will be
inaccurate. Mostly this is not a big problem.

petebacondarwin commented:
> I am in favor of fixing all mismatch example folders to match their corresponding doc folders

I will make an issue for that.
